staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,602 @@
+// a nil handler means the system should become non-quiescent.  A call
+// to Wait while the system is quiescent will be rebuffed by
+// returning `quiescent=true`.  If all the queues have no requests
+// waiting nor executing while the system is quiescent then the
+// handler will eventually be called with no locks held (even if
+// the system becomes non-quiescent between the triggering state
+// and the required call).
+//
+// The filter uses this for a priority level that has become
+// undesired, setting a handler that will cause the priority level
+// to eventually be removed from the filter if the filter still
+// wants that.  If the filter later changes its mind and wants to
+// preserve the priority level then the filter can use this to
+// cancel the handler registration.
+func (qs *queueSetImpl) Quiesce(eh fq.EmptyHandler) {
+	qs.lock.Lock()
+	defer qs.lock.Unlock()
+	if eh == nil {
+		qs.emptyHandler = eh
+		return
+	}
+	// Here we check whether there are any requests queued or executing and
+	// if not then fork an invocation of the EmptyHandler.
+	if qs.numRequestsEnqueued == 0 && qs.getRequestsExecuting() == 0 {
+		// fork an invocation of the EmptyHandler.
+		qs.counter.Add(1)
+		go func() {
+			eh.HandleEmpty()
+		}()
+	}
+
+	qs.emptyHandler = eh
+}
+
+// Wait in the happy case, shuffle shards the given request into
+// a queue and eventually dispatches the request from that queue.
+// Dispatching means to return with `quiescent==false` and
+// `execute==true`.  In one unhappy case the request is
+// immediately rebuffed with `quiescent==true` (which tells the
+// filter that there has been a timing splinter and the filter
+// re-calcuates the priority level to use); in all other cases
+// `quiescent` will be returned `false` (even if the system is
+// quiescent by then).  In the non-quiescent unhappy cases the
+// request is eventually rejected, which means to return with
+// `execute=false`.  In the happy case the caller is required to
+// invoke the returned `afterExecution` after the request is done
+// executing.  The hash value and hand size are used to do the
+// shuffle sharding.
+func (qs *queueSetImpl) Wait(hashValue uint64) (quiescent, execute bool, afterExecution func()) {
+	// TODO(aaron-prindle) verify locking is correct
+
+	qs.lockAndSyncTime()
+	// TODO(aaron-prindle) verify and test quiescent
+	// A call to Wait while the system is quiescent will be rebuffed by
+	// returning `quiescent=true`.
+	if qs.emptyHandler != nil {
+		qs.lock.Unlock()
+		return true, false, func() {}
+	}
+
+	// ========================================================================
+	// Step 1:
+	// 1) Start with shuffle sharding, to pick a queue.
+	// 2) Reject old requests that have been waiting too long
+	// 3) Reject current request if there is not enough concurrency shares and
+	// we are at max queue length
+	// 4) If not rejected, create a packet and enqueue
+	pkt := qs.timeoutOldRequestsAndRejectOrEnqueue(hashValue)
```

Comment:
s/pkt/req/
The type of this thing is `*Request`, not `*Packet`.

Reply:
updated